README.md
gitlink
================

[![Travis build status](https://travis-ci.org/colearendt/gitlink.svg?branch=master)](https://travis-ci.org/colearendt/gitlink)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/colearendt/gitlink?branch=master&svg=true)](https://ci.appveyor.com/project/colearendt/gitlink)
[![Coverage status](https://codecov.io/gh/colearendt/gitlink/branch/master/graph/badge.svg)](https://codecov.io/github/colearendt/gitlink?branch=master)

Include useful GitHub links in your Shiny applications and R Markdown
documents quickly and easily with the `gitlink` package\!

For example,

``` r
# note that this does not actually work for pure markdown documents like this README
ribbon_css("https://github.com/colearendt/gitlink")
```

![demo Shiny application with "Fork Me on Github" ribbon](man/figures/gitlink-ex.png)

### Getting Started

The package can be installed from CRAN:

```r
install.packages("gitlink")
```

Or you can install the latest development version with:

``` r
devtools::install_github("colearendt/gitlink")
```

The `ribbon_css` function is highly customizable, as you can set any color, font, font-size,
or other CSS properties with the parameters provided.

For example:

```r
  ribbon_css("mylink.com", position = "left")
  ribbon_css("mylink.com", color = "black", font_color = "white", border_color = "white")
  ribbon_css("mylink.com", text = "Fork me PLEASE", fade = FALSE)
```

For more advanced examples, see `?ribbon_css`

# Code of Conduct

Please note that the 'gitlink' project is released with a [Contributor Code of
Conduct](https://github.com/colearendt/gitlink/blob/master/.github/CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.

# Other Musings

Some interesting links that we have stumbled upon in our travels. Perhaps
interesting for future or related work:

- https://buttons.github.io/
- https://github.com/tholman/github-corners/
- http://code.danyork.com/2011/09/29/how-to-create-github-style-fork-me-ribbons-using-only-css/
- https://unindented.org/articles/github-ribbon-using-css-transforms/
- https://ghbtns.com/


NEWS.md
# gitlink 0.1.1

* Added a `NEWS.md` file to track changes to the package.
* Initial release
* Provide `ribbon_css()` with nice styling and hover behavior
* Provide limited `ribbon_img()`


tests/testthat.R
library(testthat)
library(gitlink)

test_check("gitlink")


tests/testthat/app.R
library(shiny)
library(gitlink)

ui <- fluidPage(

  # Application title
  titlePanel("Old Faithful Geyser Data"),

  ribbon_css("https://github.com/colearendt"),

  # Sidebar with a slider input for number of bins
  sidebarLayout(
    sidebarPanel(
      sliderInput(
        "bins",
        "Number of bins:",
        min = 1,
        max = 50,
        value = 30
      )
    ),

    # Show a plot of the generated distribution
    mainPanel(
      plotOutput("distPlot")
    )
  )
)

# Define server logic required to draw a histogram
server <- function(input, output) {
  output$distPlot <- renderPlot({
    # generate bins based on input$bins from ui.R
    x <- faithful[, 2]
    bins <- seq(min(x), max(x), length.out = input$bins + 1)

    # draw the histogram with the specified number of bins
    hist(x, breaks = bins, col = "darkgray", border = "white")
  })
}

# Run the application
shinyApp(ui = ui, server = server)


tests/testthat/test-ribbon.R
context("ribbon_css")

test_that("link is passed correctly", {
  res <- ribbon_css("https://thelink.example.com")
  expect_match(as.character(res), 'href="https://thelink.example.com"')
})

test_that("style is included when needed", {
  res <- ribbon_css("https://opacity.example.com", fade = TRUE)
  expect_match(as.character(res), '<div class="ribbon-parent".*>.*</div>')
  expect_match(as.character(res), "<style>.*</style>")

  res_none <- ribbon_css("https://opacity.com", fade = FALSE)
  expect_false(grepl("<style>.*</style>", as.character(res_none)))
})

test_that("text is included", {
  msg <- "Fork me Now... testing"
  res <- ribbon_css("https://text.example.com", text = msg)
  expect_match(as.character(res), paste0(">", msg, "</a>"))
})

test_that("position is configurable", {
  res_left <- ribbon_css("https://left.example.com", position = "left")
  expect_match(as.character(res_left), ";left:")
  expect_false(grepl(";right:", as.character(res_left)))

  res_right <- ribbon_css("https://right.example.com", position = "right")
  expect_match(as.character(res_right), ";right:")
  expect_false(grepl(";left:", as.character(res_right)))
})

test_that("fails on bad position", {
  expect_error(
    ribbon_css("https://invalid.example.com", position = "invalid"),
    "invalid"
  )
})

test_that("rmarkdown render works", {
  skip_if_not_installed("rmarkdown")
  skip_if_not(rmarkdown::pandoc_available())
  rmd <- rprojroot::find_testthat_root_file("rmd.Rmd")
  output <- fs::file_temp("rmd", ext = ".html")
  rmarkdown::render(rmd, output_file = output, quiet = TRUE)

  res <- readLines(output)
  expect_true(
    any(
      as.logical(
        lapply(
          res,
          function(x) {
            grep("ribbon-parent", x)
          }
        )
      )
    )
  )
})

test_that("app render works", {
  skip_if_not_installed("shiny")
  skip("test not yet implemented")
})


context("ribbon_img")

test_that("works as expected", {
  res <- ribbon_img("https://thelink.example.com")
  expect_match(as.character(res), 'href="https://thelink.example.com"')
  expect_match(as.character(res), "<img")
})

test_that("position is configurable", {
  res_left <- ribbon_img("https://left.example.com", position = "left")
  expect_match(as.character(res_left), "; left :")
  expect_false(grepl("; right :", as.character(res_left)))

  res_right <- ribbon_img("https://right.example.com", position = "right")
  expect_match(as.character(res_right), "; right :")
  expect_false(grepl("; left :", as.character(res_right)))
})

test_that("fails on bad position", {
  skip("not implemented")
  expect_error(
    ribbon_img("https://invalid.example.com", position = "invalid"),
    "invalid"
  )
})


test_that("color_lookup succeeds", {

  for (color in get_color_lookup()) {
    color_val <- get_color_lookup()[[color]]
    res_color <- ribbon_img("https://color.example.com", color = color)
    expect_match(as.character(res_color), paste(color, color_val, sep = "_"))

  }
})


R/github.R
# https://blog.github.com/2008-12-19-github-ribbons/
# http://www.webdesigndev.com/20-beautiful-free-ribbon-elements-for-your-website/

#' Build a GitHub Ribbon with CSS
#'
#' The benefits of using CSS are:
#'  - any color you can dream up
#'  - any font / font color you want
#'  - any text you want
#'  - customizable size / location / etc.
#'  - do not have to worry about DPI and image resolution
#'  - link is only clickable on the banner itself
#'
#'  CSS passed to the function will override the default CSS to
#'  give you the ultimate flexibility in customizing the ribbon to
#'  fit your needs.
#'
#'  @param link The URL that the ribbon will link to
#'  @param position The position to place the ribbon in (either "left" or "right")
#'  @param color Any css-valid color specification for the background of the ribbon
#'  @param font_color Any css-valid color specification for the text of the ribbon
#'  @param border_color Any css-valid color specification for the border of the ribbon
#'  @param text The text to show on the ribbon
#'  @param ... key=value CSS passed along to the ribbon div
#'  @param link_css A list of key=value CSS passed along to the link text
#'  @param parent_css A list of key=value CSS passed along to the parent div of the ribbon
#'
#'  @examples
#'  ribbon_css("https://github.com/colearendt/gitlink")
#'  ribbon_css("https://github.com/colearendt/gitlink", position = "left", color = "#e4e4e4")
#'  ribbon_css("https://github.com/colearendt/gitlink", position = "left", color = "#eafffc")
#' @export
ribbon_css <- function(link, position = c("left","right"), color = "white", font_color = "black", border_color = "white", text = "Fork me on GitHub", ..., link_css = list(), parent_css = list()) {
  ribbon_css <- rlang::list2(...)
  # calculate location based on position
  if (length(position) > 1) {
    position <- "right"
  }
  if (position == "right") {
    location <- list(right = "-50px", transform = "rotate(45deg)")
    parent_location <- list(right = "0px")
  } else if (position == "left") {
    location <- list(left = "-50px", transform = "rotate(-45deg)")
    parent_location <- list(left = "0px")
  } else {
    stop(sprintf("Unrecognized position: %s", position))
  }

  css_ribbon_parent <- do.call(
    htmltools::css,
    c(list(
      "position" = "absolute",
      "top" = "0px",
      "overflow" = "hidden",
      "width" = "150px",
      "height" = "150px"
      ),
    parent_location,
    parent_css
    )
  )
  css_ribbon <- do.call(
    htmltools::css,
    args = c(list(
    "background-color" = color,
    "overflow" = "hidden",
    "white-space" = "nowrap",
    "position" = "absolute",
    "top" = "45px",
    "box-shadow" = "0 0 10px #888"
    # clip-path: polygon(-50% -50%, 80% -50%, 105% 150%, -50% 150%);
    ),
    location,
    ribbon_css
    )
  )
  css_a <- do.call(
    htmltools::css,
    args = c(list(
    "border" = paste("1px solid", border_color),
    "color" = font_color,
    "display" = "block",
    "font" = "bold 95% 'Collegiate', Arial, sans-serif",
    "margin" = "1px 0",
    "padding" = "6px 50px",
    "text-align" = "center",
    "text-decoration" = "none",
    "letter-spacing" = "-0.3px"
    ),
    link_css)
    #"text-shadow" = "0 0 5px #444"
  )
  html_prep <- htmltools::div(
    class = "ribbon-parent",
    style = css_ribbon_parent,
    htmltools::div(
      class = "ribbon",
      style = css_ribbon,
      htmltools::a(href = link, text, style = css_a)
      )
  )
  return(html_prep)
}

#' @export
ribbon_img <- function(link, position = "right", color = "white") {
  hex = color_lookup[[color]]
  htmltools::a(
    href = link,
    htmltools::img(
      style = paste(
        "position: absolute; top: 0;",
        position,
        ": 0; border: 0; z-index: 500"
        ),
      src = paste0(
        "https://s3.amazonaws.com/github/ribbons/forkme_",
        position,
        "_",
        color,
        "_",
        hex,
        ".png"),
      alt = "Fork me on GitHub"
    )
  )
}

color_lookup <- list(
  red = "aa0000",
  green = "007200",
  darkblue = "121621",
  orange = "ff7600",
  gray = "6d6d6d",
  white = "ffffff"
)


R/ribbon.R
# https://blog.github.com/2008-12-19-github-ribbons/
# http://www.webdesigndev.com/20-beautiful-free-ribbon-elements-for-your-website/

# internal
# puts CSS styling related to opacity into the <head> in a <style> block
ribbon_opacity_css <- function(...) {
  hover_css <- rlang::list2(...)
  ribbon_css <- htmltools::css(
    "opacity" = "0.6",
    "transition" = c(
      "transition-property" = "opacity",
      "transition-duration" = "0s",
      "transition-timing-function" = "ease",
      "transition-delay" = "0s"
    )
  )
  ribbon_hover_css <- do.call(
    htmltools::css,
    c(
      list("opacity" = "1"),
      hover_css
    )
  )
  return(
    style = htmltools::tags$style(
      list(
        paste0(".ribbon:hover {", ribbon_hover_css, "}"),
        paste0(".ribbon {", ribbon_css, "}")
      )
    )
  )
}

#' Build a GitHub Ribbon with CSS
#'
#' @description A customizable CSS ribbon with a link to "Fork me on GitHub."
#'   CSS passed to the function will override the default CSS to give you the
#'   ultimate flexibility in customizing the ribbon to fit your needs.
#'
#' @details The benefits of using CSS are:
#'
#'  - any color you can dream up
#'
#'  - any font / font color you want
#'
#'  - any text you want
#'
#'  - customizable size / location / etc.
#'
#'  - do not have to worry about DPI and image resolution
#'
#'  - link is only clickable on the banner itself
#'
#'  - fancy fade / hover CSS out of the box
#'
#' @param link The URL that the ribbon will link to
#' @param position The position to place the ribbon in (either "left" or "right"). Default: right
#' @param color Any css-valid color specification for the background of the ribbon
#' @param font_color Any css-valid color specification for the text of the ribbon
#' @param border_color Any css-valid color specification for the border of the ribbon
#' @param text The text to show on the ribbon
#' @param ... key=value CSS passed along to the ribbon div
#' @param fade boolean. Whether or not the default opacity should be set to < 1, but transition on hover. Default TRUE
#' @param link_css A list of key=value CSS passed along to the link text
#' @param parent_css A list of key=value CSS passed along to the parent div of the ribbon
#' @param hover_css A list of key=value CSS passed along to the .ribbon:hover CSS
#'
#' @return HTML that can be injected into any output
#'
#' @examples
#' ribbon_css("https://github.com/colearendt/gitlink")
#' ribbon_css("https://github.com/colearendt/gitlink", position = "left", color = "#e4e4e4" )
#' ribbon_css("https://github.com/colearendt/gitlink", position = "left", color = "#eafffc")
#'
#' # make default opacity 1
#' ribbon_css("https://github.com/colearendt/gitlink", fade = FALSE)
#'
#' # customize the hover css
#' ribbon_css("https://github.com/colearendt/gitlink", hover_css = list("opacity" = "0.9"))
#'
#' url <- "https://github.com/colearendt/gitlink"
#'
#' # this one is particularly ugly, but proves a point
#' ribbon_css(url, parent_css = list("background-color" = "red"))
#'
#' # modify the location for a flexdashboard
#' ribbon_css(url, parent_css = list(top = "50px", "z-index" = "10"))
#' @export
ribbon_css <- function(
                       link,
                       position = c("left", "right"),
                       color = "white",
                       font_color = "black",
                       border_color = "white",
                       text = "Fork me on GitHub",
                       ...,
                       fade = TRUE,
                       link_css = list(),
                       parent_css = list(),
                       hover_css = list()) {
  ribbon_css <- rlang::list2(...)
  # calculate location based on position
  if (length(position) > 1) {
    position <- "right"
  }
  if (position == "right") {
    location <- list(right = "-50px", transform = "rotate(45deg)")
    parent_location <- list(right = "0px")
  } else if (position == "left") {
    location <- list(left = "-50px", transform = "rotate(-45deg)")
    parent_location <- list(left = "0px")
  } else {
    stop(sprintf("Unrecognized position: %s", position))
  }

  css_ribbon_parent <- do.call(
    htmltools::css,
    c(
      list(
        "position" = "absolute",
        "top" = "0px",
        "overflow" = "hidden",
        "width" = "150px",
        "height" = "150px",
        "z-index" = "5",
        "pointer-events" = "none"
      ),
      parent_location,
      parent_css
    )
  )
  css_ribbon <- do.call(
    htmltools::css,
    args = c(
      list(
        "background-color" = color,
        "overflow" = "hidden",
        "white-space" = "nowrap",
        "position" = "absolute",
        "top" = "45px",
        "box-shadow" = "0 0 10px #888",
        "pointer-events" = "auto"
        # clip-path: polygon(-50% -50%, 80% -50%, 105% 150%, -50% 150%);
      ),
      location,
      ribbon_css
    )
  )
  css_a <- do.call(
    htmltools::css,
    args = c(
      list(
        "border" = paste("1px solid", border_color),
        "color" = font_color,
        "display" = "block",
        "font" = "bold 95% 'Collegiate', Arial, sans-serif",
        "margin" = "1px 0",
        "padding" = "6px 50px",
        "text-align" = "center",
        "text-decoration" = "none",
        "letter-spacing" = "-0.3px"
      ),
      link_css
    )
    # "text-shadow" = "0 0 5px #444"
  )
  html_prep <- htmltools::div(
    class = "ribbon-parent",
    style = css_ribbon_parent,
    htmltools::div(
      class = "ribbon",
      style = css_ribbon,
      htmltools::a(href = link, text, style = css_a, target = "_blank")
    )
  )

  if (fade) {
    return(htmltools::tagList(html_prep, ribbon_opacity_css(!!! hover_css)))
  } else {
    return(html_prep)
  }
}

#' Build a GitHub Ribbon from an Image
#'
#' Publicly available ribbon images began the "GitHub ribbon craze." This is a
#' helper function that client-side retrieves these images. As a result, only
#' certain colors are available. \link{ribbon_css} is usually
#' preferred for many reasons.
#'
#' NOTE that this creates a dependency on these image files, which
#' are hosted publicly by GitHub on AWS.
#'
#' @param link The URL that the ribbon will link to
#' @param position The position to place the ribbon in (either "left" or "right")
#' @param color The color for the ribbon. One of get_color_lookup()
#'
#' @return HTML that can be injected into any HTML output
#'
#' @seealso \code{\link{ribbon_css}}
#'
#' @export
ribbon_img <- function(link, position = "right", color = "white") {
  hex <- color_lookup[[color]]
  htmltools::a(
    href = link,
    htmltools::img(
      style = paste(
        "position: absolute; top: 0;",
        position,
        ": 0; border: 0; z-index: 500"
      ),
      src = paste0(
        "https://s3.amazonaws.com/github/ribbons/forkme_",
        position,
        "_",
        color,
        "_",
        hex,
        ".png"
      ),
      alt = "Fork me on GitHub"
    )
  )
}

#' @rdname ribbon_img
#' @export
get_color_lookup <- function() {
  return(color_lookup)
}

color_lookup <- list(
  red = "aa0000",
  green = "007200",
  darkblue = "121621",
  orange = "ff7600",
  gray = "6d6d6d",
  white = "ffffff"
)
